pkg/patterns/declarative/reconciler.go

```diff
@@ -391,6 +448,17 @@ func (r *Reconciler) injectOwnerRef(ctx context.Context, instance DeclarativeObj
 	return nil
 }
 
+func (r *Reconciler) tryToSetCounter() error {
+	if mapping, err := r.mgr.GetRESTMapper().RESTMapping(r.gvk.GroupKind(), r.gvk.Version); err == nil {
```

Comment:
Possible simplification:  rather than have the caller check `setCounter`, you could check in this function for the cached `rc` and `rfc` values, and return them.

Also I think you need a mutex guarding `setCounter` because I think Reconcile can be shared between goroutines.

If you're able to refactor `rc` and `rfc` into the same object, I would just store it in a single field e.g. `reconcileMetrics`.  Then I think we can change this method to be `getReconcileMetrics() (*reconcileMetrics, error)`

Reply:
I'll create type `reconcileMetrics` in `metrics.go` and use it. And I'll stop checking whether CRD is namespaced through `RESTMapper`, because I'll leave namespace as empty value if `reconcile.Request.Namaspace` is empty.